consensus/core/src/dag_state.rs

```diff
@@ -123,28 +157,58 @@ impl DagState {
         }
     }
 
-    /// Gets a copy of an uncommitted block. Returns None if not found.
-    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
-    pub(crate) fn get_uncommitted_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
-        self.recent_blocks.get(reference).cloned()
+    /// Gets a block by checking cached recent blocks then storage.
+    /// Returns None when the block is not found.
+    pub(crate) fn get_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
+        self.get_blocks(&[*reference]).pop().unwrap()
     }
 
-    /// Gets a copy of the uncommitted blocks. Returns None for each block not found.
-    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
-    pub(crate) fn get_uncommitted_blocks(
-        &self,
-        references: Vec<BlockRef>,
-    ) -> Vec<Option<VerifiedBlock>> {
-        references
-            .into_iter()
-            .map(|reference| self.recent_blocks.get(&reference).cloned())
-            .collect()
+    /// Gets blocks by checking cached recent blocks in memory then storage.
+    /// An element is None when the corresponding block is not found.
+    pub(crate) fn get_blocks(&self, block_refs: &[BlockRef]) -> Vec<Option<VerifiedBlock>> {
+        let mut blocks = vec![None; block_refs.len()];
+        let mut missing = Vec::new();
+
+        for (index, block_ref) in block_refs.iter().enumerate() {
+            if let Some(block) = self.recent_blocks.get(block_ref) {
+                blocks[index] = Some(block.clone());
+            } else {
+                missing.push((index, block_ref));
+            }
+        }
+
+        if missing.is_empty() {
+            return blocks;
+        }
+
+        let missing_refs = missing
+            .iter()
+            .map(|(_, block_ref)| **block_ref)
+            .collect::<Vec<_>>();
+        let store_results = self
+            .store
+            .read_blocks(&missing_refs)
+            .unwrap_or_else(|e| panic!("Failed to read from storage: {:?}", e));
+        self.context
+            .metrics
+            .node_metrics
+            .dag_state_store_read_count
+            .with_label_values(&[&"get_blocks"])
+            .inc();
+
+        for ((index, _), result) in missing.into_iter().zip(store_results.into_iter()) {
+            blocks[index] = result;
+        }
+
+        blocks
     }
 
     /// Gets all uncommitted blocks in a slot.
     /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
     pub(crate) fn get_uncommitted_blocks_at_slot(&self, slot: Slot) -> Vec<VerifiedBlock> {
-        // TODO: evauluate if we should panic `if slot.round <= last_commit_round`
+        // TODO: either panic below when the slot is below the last committed round,
```

Comment:
should we assert slot is not a genesis slot?

Reply:
I think more cleanup in committer is needed to actually avoid calling this function on genesis slots. After that, we can apply the TODO and panic when reading any slot at or below last committed, including genesis slots.